### PR TITLE
Fix: Potential fix for negative statistics values + optimizations

### PR DESCRIPTION
**Pull request by GuyKh**
Solves #72

### CODE REVIEW

File: custom_components/iec/coordinator.py [hunk 1]
```diff
@@ -163,7 +168,14 @@ async def _async_update_data(
...
                 monthly_report_req_date: datetime = TIMEZONE.localize(datetime.today().replace(hour=1, minute=0,
                                                                                                second=0, microsecond=0))
-                devices = await self.api.get_devices(contract_id)
+
+                # reduce number of calls
+                if contract_id in self._devices_by_contract_id:
+                    devices = self._devices_by_contract_id[contract_id]
+                else:
+                    devices = await self.api.get_devices(str(contract_id))
+                    self._devices_by_contract_id[contract_id] = devices
```
Comment (myakove) [inline custom_components/iec/coordinator.py:177]:
```suggestion
                devices = self._devices_by_contract_id.get(contract_id)
                if not devices:
                    devices = await self.api.get_devices(str(contract_id))
                    self._devices_by_contract_id[contract_id] = devices
```

File: custom_components/iec/coordinator.py [hunk 2]
```diff
@@ -247,10 +261,16 @@ async def _insert_statistics(self, contract_id: int, is_smart_meter: bool) -> No
             return
 
         _LOGGER.debug(f"Updating statistics for IEC Contract {contract_id}")
-        devices = await self.api.get_devices(str(contract_id))
-        month_ago_time = (datetime.now() - timedelta(weeks=4))
+        if contract_id in self._devices_by_contract_id:
+            devices = self._devices_by_contract_id[contract_id]
+        else:
+            devices = await self.api.get_devices(str(contract_id))
+            self._devices_by_contract_id[contract_id] = devices
```
Comment (myakove) [inline custom_components/iec/coordinator.py:268]: Repeated code, I suggest extracting it to function (will be easier to fix issue when needed in one place only)

File: custom_components/iec/coordinator.py [hunk 3]
```diff
@@ -247,10 +261,16 @@ async def _insert_statistics(self, contract_id: int, is_smart_meter: bool) -> No
...
 
-        kwh_price = await self.api.get_kwh_tariff() / 100
+        if not self._kwh_tariff:
+            self._kwh_tariff = await self.api.get_kwh_tariff() / 100
+
+        kwh_price = self._kwh_tariff
```
Comment (myakove) [inline custom_components/iec/coordinator.py:273]: ditto

**Comment by myakove** [comment]
I hope it's OK that I reviewed the PR :)

**Comment by GuyKh** [comment]
@myakove - take another look

**Pull request merged**